pkg/networkpath/traceroute/tcp/utils.go

```diff
@@ -0,0 +1,423 @@
+	pseudoHeader = append(pseudoHeader, 0) // reserved
+	pseudoHeader = append(pseudoHeader, byte(ipHdr.Protocol))
+	pseudoHeader = append(pseudoHeader, 0, byte(len(tcpHeader))) // tcp length
+
+	return checksum(append(pseudoHeader, tcpHeader...))
+}
+
+func listenAnyPacket(icmpConn *ipv4.RawConn, tcpConn *ipv4.RawConn, timeout time.Duration, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, seqNum uint32) (net.IP, uint16, layers.ICMPv4TypeCode, time.Time, error) {
+	var tcpErr error
+	var icmpErr error
+	var wg sync.WaitGroup
+	var icmpIP net.IP
+	var tcpIP net.IP
+	var icmpCode layers.ICMPv4TypeCode
+	var port uint16
+	wg.Add(2)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	go func() {
+		defer wg.Done()
+		defer cancel()
+		tcpIP, port, _, tcpErr = handlePackets(ctx, tcpConn, "tcp", localIP, localPort, remoteIP, remotePort, seqNum)
+	}()
+	go func() {
+		defer wg.Done()
+		defer cancel()
+		icmpIP, _, icmpCode, icmpErr = handlePackets(ctx, icmpConn, "icmp", localIP, localPort, remoteIP, remotePort, seqNum)
+	}()
+	wg.Wait()
+	// TODO: while this is okay, we
+	// should do this more cleanly
+	finished := time.Now()
+
+	if tcpErr != nil && icmpErr != nil {
+		_, tcpCanceled := tcpErr.(canceledError)
+		_, icmpCanceled := icmpErr.(canceledError)
+		if icmpCanceled && tcpCanceled {
+			// TODO: better handling of listener timeout case
+			// this signifies an unknown hop
+			//
+			// TODO: better handling of the mismatch case which
+			// right now becomes an unknown hop
+			//
+			// For now, return nil error with empty hop data
+			log.Debug("timed out waiting for responses")
+			return net.IP{}, 0, 0, finished, nil
+		}
+		log.Errorf("TCP Error: %s", tcpErr.Error())
+		log.Errorf("ICMP Error: %s", icmpErr.Error())
+		return net.IP{}, 0, 0, finished, multierr.Append(fmt.Errorf("tcp error: %w", tcpErr), fmt.Errorf("icmp error: %w", icmpErr))
+	}
+
+	// if there was an error for TCP, but not
+	// ICMP, return the ICMP response
+	if tcpErr != nil {
+		return icmpIP, port, icmpCode, finished, nil
+	}
+
+	// return the TCP response
+	return tcpIP, port, 0, finished, nil
+}
+
+// handlePackets in its current implementation should listen for the first matching
+// packet on the connection and then return. If no packet is received within the
+// timeout, it should return a timeout exceeded error
+func handlePackets(ctx context.Context, conn *ipv4.RawConn, listener string, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, seqNum uint32) (net.IP, uint16, layers.ICMPv4TypeCode, error) {
+	buf := make([]byte, 1024)
+	for {
+		select {
+		case <-ctx.Done():
+			return net.IP{}, 0, 0, canceledError("listener canceled")
+		default:
+		}
+		now := time.Now()
+		err := conn.SetReadDeadline(now.Add(time.Millisecond * 100))
+		if err != nil {
+			return net.IP{}, 0, 0, fmt.Errorf("failed to read: %w", err)
+		}
+		header, packet, _, err := conn.ReadFrom(buf)
```

Comment:
Have you looked at https://pkg.go.dev/github.com/google/gopacket#hdr-Fast_Decoding_With_DecodingLayerParser for parsing the packet?

Reply:
I did, I found it a bit after I originally implemented this. I'm going to have to do it in 2 steps for ICMP since we sometimes have to extend the inner TCP header when the ICMP response doesn't contain the whole thing. There's a future iteration of this where it might make more sense to just do unmarshal manually